docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+  zero or more instances of objects of type `X` which have been registered,
+  using a `compositor` function provided as a constructor argument (or via
+  a subsequent `compose` call.)
+  * `register(factory : (function () : X), [options] : RegistrationOptions)`
+    registers an object (as returned by the provided `factory` function)
+    with this provider. Evaluation of the provided `factory` will be
+    deferred until the first `get()` call to the `Provider`.
+  * `compose(compositor : (function (X[]) : S), [options] : RegistrationOptions)`
+    introduces a new strategy for converting an array of registered objects
+    of type `X` into a single instance of an object of type `S`. The
+    highest-priority `compositor` that has been registered in this fashion
+    will be used to assemble the provided object (before decoration)
+* `ServiceProvider<S>` extends `CompositeProvider<S, S>` and provides analogous support
+  for the _composite services_
+  pattern used throughout Open MCT (which, in turn, is a superset of the
+  functionality needed for plain services.)
+* `Registry<T>` extends `CompositeProvider<T, T[]>` and provides analogous support for
+  _extension categories_, also
+  used ubiquitously through Open MCT.
+* `RegistrationOptions` provides a location for additional
+  registration-related metadata to be provided.
+  * `priority` is used to control the ordering and/or preference of
+    factories, compositors, and/or decorators,
+    [as in the existing Framework Layer.](http://nasa.github.io/openmctweb/guide/#priority)
+
+# Examples
+
+The following examples are provided to illustrate the intended usage of
+the Registration API. Particular attention is given to obeying and
+utilizing the "dependency injection as code style"
+[decision from the API Redesign](APIRedesign.md#decisions).
+
+## Building Applications on Open MCT
+
+Applications built using Open MCT are expected to extend the `MCT` base
+class and should typically self-install distinguishing plugins during
+the constructor call. Any pre-installed plugins that are undesirable
+should also be uninstalled at this point (to support such usage,
+`MCT` should expose instances of any installed plugins that are
+considered optional.)
+
+For example:
+
+```
+define(['mct', './plugins'], function (mct, plugins) {
+  var MCT = mct.MCT;
+
+  function Variant() {
+    MCT.call(this);
+
+    this.install(new plugins.SomePlugin());
+    this.install(new plugins.SomeOtherPlugin(this.core));
+
+    this.uninstall(this.plugins.persistence.localStorage);
+  }
+
+  Variant.prototype = Object.create(MCT.prototype);
+
+  return Variant;
+});
+```
+
+Running an application build using Open MCT then typically looks like:
+
+```
+define(['./Variant], function (Variant) {
+  new Variant().run();
+});
+```
+
+## Writing Plugins
+
+A plugin for Open MCT should inherit from the `Plugin` base class.
+
+Plugins will typically use extension points exposed by other plugins;
```

Comment:
I see two useful things that I can't do like this:

1. Modularize my extension points in groups, to clarify what plugins depend on. This goes back to [discussion above](https://github.com/nasa/openmct/pull/705#discussion_r56413247)
2. Implement extensible features via plugin. As a concrete example, one might have a Data Product plugin which shows previews of certain known data product types. A new mission has new Data Products? Okay, add a plugin which knows how to preview those.

I don't think Open MCT _needs_ to provide support for declaring new extension points, but at minimum we should be compatible (without weird workarounds.) But since we need a useful mechanism for declaring extension points, why not expose this to allow more to be declared? Isn't it better if non-MCT extension points utilize the same interfaces and patterns as MCT's own?

That said, I do agree that plugin instances are not an ideal place for these. It loads a lot of responsibility on plugins - declare, use, and configure extension points. Declaration, at least, could be moved out (leaving an open question of "to what?")

> My concern is that we confuse the internal architecture of OpenMCT with the plugin architecture that is used to customize it for different purposes.

What do we gain by having two architectures (internal versus plugin) instead of one? Sounds like we design/implement/maintain twice as much, and exercise each one half as much. I'd prefer to solve both the same way; there are clear benefits and no inherent problems that I see.